README
This script will help you prepare a server for cPanel installation. It will check various requirements and let you know if there is anything that needs to be fixed and how to do so before proceeding with the install. 

To run this script directly run the following on your command line. If you have issues grabbing this with ssl use "--insecure" after "curl -s"

curl -s -L go.cpanel.net/preinstallcheck | sh

To run this script directly with options use the following. To find a full list of options use --help

bash <(curl -s -L go.cpanel.net/preinstallcheck) --option

To run this script locally on the server run the following

wget go.cpanel.net/preinstallcheck
sh preinstallcheck
or with options
sh preinstallcheck --option


CHANGELOG
1.4.2:
- Added colors for [INFO] output.

1.4.1:
- Added NetworkManager Detection.

1.4.0:
- Tidied up the code
- Added Amazon Linux as a Supported OS
- Finally noted that RHEL 7 based OS's are fully supported
- Removed Support for installing on CentOS 5 and 32-bit systems

1.3.1:
- Fixed Hostname Check if Dashes were in hostname
- Fixed Install Switch
- Implemented a check to see if a RHEL based OS was installed or bail
- Implemented some new checks and flags for RHEL 7 as we now support CentOS 7 and 3.x kernels in 11.50 as a Tech Preview.

1.3.0.06:
- Completely rewritten in Perl
- Imporoved and Cleaned Up Output
- It will now quit if a cPanel Install/Removal was detected and no further checks will be performed.
- There is built in support for installing perl if found to be missing at the start of the script (script starts in bash and moves to perl once verified to exist.)
- Interactivity Added. Now you will be prompted to auto fix certain issues at the end of the script, such as hostname, dig, wget, yum groups. Everything else gives you manual commands to fix.
- The Yum Group Check no longer prints out a combined PASS/FAIL with all groups. It now only prints out groups with a FAIL that need to be removed or will print a simple PASS if no FAILs found.
- Script Supports RHEL 5, 6 & 7.

1.2.10.3 - Removed the double brackets in the memory check as they appeared to be breaking this on some systems
1.2.10.2 - Fixed memory check from 500 to 512
1.2.10.1 - Added check to print out disk space
1.2.10 - Added check for courier email server, run level check, tmp 1777 perm check, minimum memory reqt check
1.2.9.5 - Install Switch now outputs the exact OS when RHEL/CENTOS/CL 5 detected.
1.2.9.4 - Fixed output in install swtich to rename CentOS 5 to RHEL 5 based machine to be more generic.
1.2.9.3 - Added mv latest to install switch before downloading latest file.
1.2.9.2 - Added warning if using an invalid option.
1.2.9.1 - Added install switch and force-install swtich for centos 5 along with warnings for using cent 5.
1.2.9 - Added switches for help menu, to disable color, improved PASS/FAIL output and added WARN output for certain sections. Changed the section output from yellow to white.  Added fix for latest perl version check in RHEL/Cent/CL 5.
1.2.8.2 - Fixed spacing of if statements to improve readability of the code
1.2.8 - Reduced spawned processes and subshells of the script by about half
1.2.7.5 - fixed bug in kernel check syntax (master branch) only updated ver # in dev branch
1.2.7.4 - additional changes to hostname check output
1.2.7.3 - Added additional output/info for fixing hostname/fqdn issues
1.2.7.2 - Added some additional checks/output to cpanel install check
1.2.7.1 - Corrected bug in cpanel install chek output
1.2.7 - Added Check for Current/Previous cPanel Installation
1.2.6 - Improved the process for selinux checks
1.2.5.2 - Fixed a regression in the code structure applied with update 1.2.5 for selinux that was fixed in 1.2.4.1
1.2.5.1 - fixed wget check
1.2.5 - Added a check for wget
1.2.4.1 - Cleaned up code structure for selinux nested if statments
1.2.4 - Removed Firewall Check
1.2.3 - Fixed FQDN Check output for when dig not installed/added check for dig.
1.2.2.6 - tweaked os/kernel check section name
1.2.2.5 - fixed bug in grsec kerel check/
1.2.2.4 - fixed issue in grsec kerel check/added additional check
1.2.2.3 - fixed path in kernel check
1.2.2.2 - added grsec to kernel check
1.2.2.1 - added color to kernel check
1.2.2.0 - Added kernel check
1.2.1.5 - fixed bug perl ver check rhel 5, removed /dev/null from perl version check
1.2.1.4 - fixed perl version check for rhel 5 OS's
1.2.1.3 - fixed group check rhel 5 mail server
1.2.1.2 - Fixed a bug in output
1.2.1.1 - fixed names on a few yum groups
1.2.1.0 - Group check updated to safer method
1.2.0.7 - changed perl installed version check
1.2.0.6 - added additional info checks for perl
1.2.0.5 - switched output to a pass/fail system
1.2.0.4 - more code cleanup
1.2.0.3 - cleaned up os check output
1.2.0.2 - added space to \_ hostname
1.2.0.1 - cleaned up os and hostname check output
1.2 - Cleaned up Output, Added Color for easier readability, added perl check
1.1.2.04 - additional code cleanup
1.1.2.03 - Added FQDN/Hostname check, cleaned up output for yum group 

preinstall_check.sh
#!/bin/bash
###############################
##  cPanel Preinstall Check  ##
##  Version 1.1.1            ##
##  By: Matthew Vetter       ##
##      cPanel, Inc.         ##
###############################

#Check SELINUX Status

echo "=====SELINUX CHECK=====";

selinuxfile="/etc/selinux/config"

if [ -f "$selinuxfile" ] ; then
if ``cat "$selinuxfile" | grep "#SELINUX=" > /dev/null`` ; then
echo "SELINUX is commented out! You need to uncomment this (remove the # from in front of SELINUX) and set this to disabled!";
echo "To fix this please review the following article http://www.cyberciti.biz/faq/howto-turn-off-selinux/ and apply the permanent fix by editing /etc/sysconfig/selinux and rebooting the server";
elif [ -f "$selinuxfile" ] ; then
if ``cat "$selinuxfile" | grep "SELINUX=" | grep "enforcing" > /dev/null`` ; then
echo "SELINUX is set to enforcing! You need to set this to disabled!";
echo "To fix this please review the following article http://www.cyberciti.biz/faq/howto-turn-off-selinux/ and apply the permanent fix by editing /etc/sysconfig/selinux and rebooting the server";
elif [ -f "$selinuxfile" ] ; then
if ``cat "$selinuxfile" | grep "SELINUX=" | grep "permissive" > /dev/null`` ; then
echo "SELINUX is set to permissive! You need to set this to disabled!";
echo "To fix this please review the following article http://www.cyberciti.biz/faq/howto-turn-off-selinux/ and apply the permanent fix by editing /etc/sysconfig/selinux and rebooting the server";
else
echo "Nothing to Fix. SELINUX appears to be disabled already!"
fi
fi
fi
fi

echo "=====FIREWALL CHECK=====";

# Check if iptable disabled in chkconfig
if ``chkconfig --list | grep iptables | grep "0:off	1:off	2:off	3:off	4:off	5:off	6:off" > /dev/null`` ; then
echo "Firewall off in ChkConfig. Nothing to Fix!";
elif ``chkconfig --list | grep iptables | grep "on" > /dev/null`` ; then
echo "Firewall enabled in ChkConfig. You should turn this off.";
echo "To turn this off run: chkconfig iptables off";
fi

# Check if Firewall Running
if ``/etc/init.d/iptables status | grep "Table: filter" > /dev/null`` ; then
echo "Firewall Running. You should disable this.";
echo "To disable this run: /etc/init.d/iptables save; /etc/init.d/iptables stop";
elif ``/etc/init.d/iptables status | grep "Firewall is not running" > /dev/null`` ; then
echo "Firewall Not Running. Nothing to Fix!";
fi

#Check if Perl Installed

echo "=====PERL CHECK=====";

if perl < /dev/null > /dev/null 2>&1  ; then
echo "Perl is Installed. Nothing to Fix!"
else
echo "Perl not Installed. You need to install this!";
echo "To install perl run: yum install perl";
fi

# Check for Yum Groups All Versions (Group names the same accross all versions)

echo "=====YUM GROUPS CHECK=====";

if ``echo "n" | yum groupremove "FTP Server" | grep "Removing:" > /dev/null`` ; then
echo "FTP Server is Installed. You should remove this";
echo 'To remove this run: yum groupremove "FTP Server"';
elif ``echo "n" | yum groupremove "FTP Server" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "FTP Server is Not Installed";
fi

if ``echo "n" | yum groupremove "Web Server" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "Web Server is Installed. You should remove this";
echo 'To remove this run: yum groupremove "Web Server"';
elif ``echo "n" | yum groupremove "Web Server" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "Web Server is Not Installed";
fi

if ``echo "n" | yum groupremove "X Window System" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "X Window System is Installed. You should remove this";
echo 'To remove this run: yum groupremove "X Window System"';
elif ``echo "n" | yum groupremove "X Window System" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "X Window System is Not Installed";
fi

# New Group Names CentOS/RHEL 6.*
if ``cat /etc/redhat-release | grep "release 6.*" > /dev/null``  ; then

if ``echo "n" | yum groupremove "E-mail Server" | grep "Removing:"  > /dev/null`` ; then
echo "=========="
echo "E-mail Server is Installed. You should remove this";
echo 'To remove this run: yum groupremove "E-mail Server"';
elif ``echo "n" | yum groupremove "E-mail Server" | grep "No packages to remove from groups"  > /dev/null`` ; then
echo "=========="
echo "E-Mail Server is Not Installed";
fi

if ``echo "n" | yum groupremove "KDE Desktop" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "KDE Desktop is Installed. You should remove this";
echo 'To remove this run: yum groupremove "KDE Desktop"';
elif ``echo "n" | yum groupremove "KDE Desktop" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "KDE Desktop is Not Installed";
fi

if ``echo "n" | yum groupremove "Desktop" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "Gnome Desktop is Installed. You should remove this";
echo 'To remove this run: yum groupremove "Desktop"';
elif ``echo "n" | yum groupremove "Desktop" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "Gnome Desktop is Not Installed";
fi

fi

# Deprecated Group Names in CentOS/RHEL 6. Will check if on CentOS/RHEL 5.*

if ``cat /etc/redhat-release | grep "release 5.*" > /dev/null``  ; then

if ``echo "n" | yum groupremove "Mail Server" | grep "Removing:" | grep -v "No group named" > /dev/null`` ; then
echo "=========="
echo "Mail Server is Installed. You should remove this";
echo 'To remove this run: yum groupremove "Mail Server"';
elif ``echo "n" | yum groupremove "Mail Server" | grep "No packages to remove from groups" | grep -v "No group named"  > /dev/null`` ; then
echo "=========="
echo "Mail Server is Not Installed";
fi

if ``echo "n" | yum groupremove "GNOME Desktop Environment" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "GNOME Desktop Environment is Installed. You should remove this";
echo 'To remove this run: yum groupremove "GNOME Desktop Environment"';
elif ``echo "n" | yum groupremove "GNOME Desktop Environment" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "GNOME Desktop Environment is Not Installed";
fi

if ``echo "n" | yum groupremove "KDE (K Desktop Environment)" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "KDE (K Desktop Environment) is Installed. You should remove this";
echo 'To remove this run: yum groupremove "KDE (K Desktop Environment)"';
elif ``echo "n" | yum groupremove "KDE (K Desktop Environment)" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "KDE (K Desktop Environment) is Not Installed";
fi

if ``echo "n" | yum groupremove "Mono" | grep "Removing:" > /dev/null`` ; then
echo "=========="
echo "Mono is Installed. You should remove this";
echo 'To remove this run: yum groupremove "Mono"';
elif ``echo "n" | yum groupremove "Mono" | grep "No packages to remove from groups" > /dev/null`` ; then
echo "=========="
echo "Mono is Not Installed";
fi

fi



cpanel_preinstall_check.sh
eval 'if [ ! -e /usr/bin/perl ] ; then yum install -y perl > /dev/null; echo "Perl was missing, so we installed it."; fi; if [ -x /usr/bin/perl ]; then exec /usr/bin/perl -x $0 ${1+"$@"}; fi;'
if 0;

#!/usr/bin/perl
###############################
##  cPanel Preinstall Check  ##
##  Version 1.4.2            ##
##  By: Matthew Vetter       ##
###############################

use Getopt::Long;

my $green  = "\e[0;32m";
my $red    = "\e[0;31m";
my $yellow = "\e[0;33m";
my $cyan   = "\e[0;36m";
my $NC     = "\e[0m";
my $UL     = "\e[4m";
my $UE     = "\e[0m";

#Let's make sure /etc/redhat-release exists before proceeding
if ( -e "/etc/redhat-release" || -e "/etc/system-release") {
    $rhelexists = "yes";
}
else {
    print "${yellow}[WARN]\t \\_ /etc/redhat-release is missing. Are you running CentOS or RedHat?${NC}\n";
    print "\t \\_ To run cPanel and this script you must be running CentOS or RedHat Enterprise Linux!\n";
    exit 0;
}

#Don't Move, Else We Break the Install Switches.
my $rhel5  = `grep "release 5.*" /etc/redhat-release`;
my $rhel6  = `grep "release 6.*" /etc/redhat-release`;
my $rhel7  = `grep "release 7.*" /etc/redhat-release`;
my $amazon = `grep -i "Amazon" /etc/os-release`;
my $bit32  = `uname -a | grep "i686\|i386"`;
chomp( my $rhel      = `awk '{print \$1, \$3}' /etc/redhat-release` );
chomp( my $rhel7prnt = `awk '{print \$1, \$4}' /etc/redhat-release` );
chomp( my $rhelrec   = `awk '{print \$1}' /etc/redhat-release` );

print "========================================================================================================================\n";

my $nocolor      = 0;
my $fixit        = 0;
my $help         = 0;
my $install      = 0;

GetOptions(
    'nocolor'       => sub { $nocolor      = 1 },
    'fix|okdoit'    => sub { $fixit        = 1 },
    'help|h|option' => sub { $help         = 1 },
    'install'       => sub { $install      = 1 }
);

if ( $help == 1 ) {
    print "-h / --help : Displays this screen\n";
    print "--nocolor : Runs Script without Color Output\n";
    print "--install : Grabs the cPanel Installer and Runs it.\n";
    print "--fix: Fix issues detected in checks.\n";
    print "========================================================================================================================\n";
    exit 0;
}
if ( $nocolor == 1 ) {
    $green  = "\e[0m";
    $red    = "\e[0m";
    $yellow = "\e[0m";
    $cyan   = "\e[0m";
    $NC     = "\e[0m";
}
if ( $install == 1 ) {
    if ( $rhel5 || $bit32 ) {
        print "${red}[FAIL] * $rhel (Linux) / 32-bit OS detected!${NC}\n";
        print "${cyan}[INFO] * cPanel no longer supports RHEL 5 or 32 bits systems. Please upgrade to CentOS 6 64-bit or CentOS 7 64-bit.${NC}\n";
    }
    elsif ( $rhel6 || rhel7 ) {
        &cpinstall;
    }
    print "========================================================================================================================\n";

    exit 0;
}

print "cPanel Preinstall Check\n";
print "[Version] 1.4.2\n";
print "[Updated] June 5 2018\n";
print "${cyan}[INFO] * This script has been deprecated and updates may no longer occur. Please utilize the cPanel Installer.${NC}\n";
print "${cyan}[INFO] * You can run the installer from this script using --install or download it manually.${NC}\n";

cpanelchk();

#resolvechk();

chomp( my $perlrhel6   = `yum info perl 2>/dev/null | awk '/Version/{print \$3}'` );
chomp( my $perlrhel5   = `yum info perl 2>/dev/null | awk '/Installed Packages/ {flag=1;next} /Available Packages/{flag=0} flag {print}'  | awk '/Version/{print \$3}'` );
chomp( my $selinuxmode = `sestatus | awk '/Current mode:/ {print \$3}'` );
chomp( my $cpversion   = `[ -e /usr/local/cpanel/version ] && awk '{print \$1}' /usr/local/cpanel/version` );
chomp( my $nmcontrolled = `systemctl status NetworkManager | grep -ow running`);

#perlchk();
wgetchk();
selinuxchk();
nmcontrolledchk();
yumgroupchk();
hostnamechk();
oskernelchk();

if ($failed) {
    if ( !$fixit ) {
        print "\n${cyan}[INFO] * YOU CAN FIX FAILURES/WARNINGS BY RE-RUNNING THE SCRIPT WITH --fix${NC}\n";
    }
}

print "\n========================================================================================================================\n";

#Sub Routines
sub cpinstall {
    if (`wget --help`) {
        chdir "/root/" or die "Cannot change to root\n";
        rename "latest" => "latest.bak";
        system("wget http://httpupdate.cpanel.net/latest");
        system("sh latest");
    }
    else {
        print "wget is not installed, cannot run installer. Please install wget with 'yum install wget'\n";
    }
}

sub cpanelchk {
    print "\n${UL}cPanel Install CHECK:${UE}\n";
    if ( -e "/usr/local/cpanel" || -e "/var/cpanel" || -e "/etc/init.d/cpanel" ) {
        print "${red}[FAIL] * cPanel Installation Found${NC}\n";
        if ( -e "/usr/local/cpanel" ) {
            print "\t \\_ cPanel/WHM $cpversion Install Detected\n";
            print "\t \\_ You will need to start with a fresh OS install / reinstall the OS before installing cPanel\n";
        }
        else {
            print "\t \_ Previous cPanel Installation/Removal Detected\n";
            print "\t \_ You will need to start with a fresh OS install / reinstall the OS before installing cPanel\n";
        }
        print "========================================================================================================================\n";
        exit 0;
    }
    else {
        print "${green}[PASS] * cPanel not Installed${NC}\n";

    }
}

#No Longer Output
sub perlchk {
    print "\n${UL}PERL CHECK:${UE}\n";
    print "${green}[PASS] * Perl Installed${NC}\n";
    if ($rhel6) {
        print "\t \\_ Latest version in Yum is - Perl $perlrhel6\n";
    }
    elsif ($rhel5) {
        print "\t \\_ Latest version in Yum is - Perl $perlrhel5\n";
    }
    else {
        print "${red}[FAIL] * Perl is not installed.${NC}\n";
        print "\t \\_ To install perl run: yum install perl\n";
    }
}

sub wgetchk {
    print "\n${UL}WGET CHECK:${UE}\n";
    if (`wget --help`) {
        print "${green}[PASS] * wget is Installed${NC}\n";
        $wgetpass = "yes";
    }
    else {
        print "${red}[FAIL] * wget not Installed${NC}\n";
        $failed = "yes";
    }

    if ( $fixit == 1 ) {
        if ( !$wgetpass ) {
            print "Would you like to fix this? Y or N: ";
            my $answer = <STDIN>;
            chomp($answer);
            if ( $answer =~ m/^y$|^yes$/i ) {
                system("yum -y install wget > /dev/null");
                if (`wget --help`) {
                    print "\t \\_ wget was installed succesfully\n";
                }
                else {
                    print "\t \\_ wget was not installed succesfully. To fix manually install with: yum install wget\n";
                }
            }
            else {
                print "\t \\_ To fix manually install wget with: yum install wget\n";
            }
        }
    }
}

sub nmcontrolledchk {
    print "\n${UL}NETWORK MANAGER CHECK:${UE}\n";
    if ($nmcontrolled) {
        print "${red}[FAIL] * NetworkManager is running${NC}\n";
        if ( $fixit == 1 ) {
            print "\t \\_ To fix this see - https://documentation.cpanel.net/display/CKB/How+to+Disable+Network+Manager\n";
        }
    }
    elsif (-e "/etc/systemd/system/dbus-org.freedesktop.NetworkManager.service") {
        print "${yellow}[WARN] * NetworkManager is not running but is enabled${NC}\n";
            if ( $fixit == 1 ) {
                 print "\t \\_ To fix this see - https://documentation.cpanel.net/display/CKB/How+to+Disable+Network+Manager\n";
             }
        }
    else {
        print "${green}[PASS] * Network Manager is disabled${NC}\n";
    }
}

sub selinuxchk {
    print "\n${UL}SELINUX  CHECK:${UE}\n";

    if (`grep "#SELINUX=" /etc/selinux/config`) {
        print "${red}[FAIL] * SELINUX${NC}\n";
        print "\t \\_ SELINUX is commented out!\n";
        if ( $fixit == 1 ) {
            print "\t \\_ To fix this edit /etc/sysconfig/selinux, uncomment SELINUX= (remove the # from in front of SELINUX), set it to disabled and then reboot the server\n";
        }
    }
    elsif (`sestatus | grep "enabled"`) {
        print "${red}[FAIL] * SELINUX${NC}\n";
        print "\t \\_ Selinux is Enabled and is currently set to $selinuxmode\n";

        if (`sestatus | grep "Mode from config file:" | grep "disabled"`) {
            print "\t \\_ However Selinux is set to disabled in the config file. The server needs to be rebooted to apply the change\n";

        }
        else {
            if ( $fixit == 1 ) {
                print "\t \\_ To fix this edit /etc/sysconfig/selinux, set SELINUX= to disabled and then reboot the server\n";
            }
        }
    }

    elsif (`sestatus | grep "disabled"`) {
        print "${green}[PASS] * SELINUX${NC}\n";
        print "\t \\_ SELINUX appears to be disabled already\n";
    }

}

sub resolvechk {
    print "\n${UL}RESOLVER CHECK:${UE}\n";
    eval {
        local $SIG{ALRM} = sub { die "alarm\n" };
        alarm 3;
        $test = `curl google.com 2>/dev/null`;
        alarm 0;
    };

    my $nameserverchk = `awk '\$1 == "nameserver" { print \$2 }' /etc/resolv.conf 2>/dev/null`;

    if ( $nameserverchk eq "" || $@ ) {
        print "${yellow}[WARN] * Resolvers Look To Be Broken! Cannot proceed until fixed!${NC}\n";
        print "\t \\_ To fix this edit /etc/resolv.conf and verify you have working resolvers listed and they are not commented with #.\n";
        print "\t \\_ Examples of googles publicly working resolvers are below:\n";
        print "\t \\_ nameserver 8.8.8.8\n \t \\_ nameserver 8.8.4.4\n";
        exit 0;
    }
    else {
        print "${green}[PASS] * Resolvers Look To Be Working${NC}\n";
    }

}

sub yumgroupchk {
    print "\n${UL}YUM GROUP/PACKAGE CHECK:${UE}\n";

    my %bad_groups = (
        'FTP server'                  => 1,
        'Web Server'                  => 1,
        'X Window System'             => 1,
        'E-mail server'               => 1,
        'KDE Desktop'                 => 1,
        'Desktop'                     => 1,
        'Mail Server'                 => 1,
        'GNOME Desktop Environment'   => 1,
        'KDE (K Desktop Environment)' => 1,
        'Mono'                        => 1,
    );

    my @installed_bad;

    open( my $output, "-|", "yum grouplist 2>/dev/null" ) or die "Problem with yum! $1";

    my $print = 0;

    while (<$output>) {

        my $group = $_;
        $group =~ s/^\s+|\s+$//g;

        if ( $group =~ /Available Groups/ ) {
            $print = 0;
        }

        if ( $bad_groups{$group} ) {
            push( @installed_bad, $group ) if $print;
        }

        if ( $group =~ /Installed Groups/ ) {
            $print = 1;
        }
    }

    close $output;

    my $courier = `rpm -qa | grep -i courier`;
    if ( $courier || @installed_bad ) {
        if ($courier) {
            print "${red}[FAIL] * \"Courier E-mail Server\" Found${NC}\n";
            $failed = "yes";
        }
        if (@installed_bad) {
            foreach my $bad (@installed_bad) {
                print "${red}[FAIL] * \"$bad\" Group Found${NC}\n";
                $failed = "yes";
            }
        }
    }
    else {
        print "${green}[PASS] * No Yum Groups/Packages Need to Be Removed${NC}\n";
        $yumpass = "yes";
    }

    if ( $fixit == 1 ) {
        if ( !$yumpass ) {
            print "Would you like to fix this? Y or N: ";
            my $answer = <STDIN>;
            chomp($answer);
            if ( $answer =~ m/^y$|^yes$/i ) {
                foreach my $bad (@installed_bad) {
                    my $pid = open( my $out, "-|", "yum groupremove -y '$bad' 2>/dev/null" ) or die "Can't fork: $!";
                    if ($pid) {
                        waitpid $pid, 0;
                        close $out;
                    }
                    print "\t \\_ Yum Group $bad was removed.\n";
                }

                if ($courier) {
                    system("rpm -e --nodeps `rpm -qa | grep -i courier`");
                    if ( `rpm -qa | grep -i courier` eq "" ) {
                        print "\t \\_ Courier has been removed succesfully.\n";
                    }
                    else {
                        print "\t \\_ Courier does not appear to have been removed\n";
                    }

                }
            }
            else {
                foreach my $bad (@installed_bad) {
                    print "\t \\_ To remove \"$bad\" group manually run: yum groupremove \"$bad\"\n";
                }
                if ($courier) {
                    print "\t \\_ To remove Courier manually run: rpm -e --nodeps `rpm -qa | grep -i courier`${NC}\n";
                }
            }
        }
    }

}

sub hostnamechk {

    print "\n${UL}HOSTNAME CHECK:${UE}\n";
    chomp( my $hostname   = `hostname` );
    chomp( my $hostnameip = `curl -s -L http://cpanel.net/myip 2>/dev/null` );
    chomp( my $fqdnhost   = `hostname | grep -P '[a-zA-Z0-9-]+[.]+[a-zA-Z0-9-]+[.]+[a-zA-Z0-9]+'` );

    if ( $hostname eq $fqdnhost ) {
        print "${green}[PASS] * The server's hostname of $hostname is a FQDN${NC}\n";
        $hostpass = "yes";
    }
    else {
        print "${red}[FAIL] * The server's hostname of $hostname is not a FQDN${NC}\n";
        $failed = "yes";
    }

    if ( $fixit == 1 ) {
        if ( !$hostpass ) {
            chomp( my $oldhostname = `hostname` );
            print "Would you like to fix this? Y or N: ";
            my $answer = <STDIN>;
            chomp($answer);
            if ( $answer =~ m/^y$|^yes$/i ) {
                my $hostnamefile = '/etc/sysconfig/network';
                print "Enter new hostname\n";
                print "\t \\_ For example a FQDN should look like server.domain.com or server.domain.co.uk\n";
                my $newhostname = <STDIN>;
                chomp($newhostname);

                if ( $rhel5 || $rhel6 ) {

                    system("hostname $newhostname");
                    my $hostsource = '/etc/sysconfig/network';
                    my $hosttemp   = '/etc/sysconfig/network.new';
                    chomp( my $date = `date +%Y-%m-%d-%H:%M:%S` );
                    my $hostbak = '/etc/sysconfig/network.bak.' . $date;

                    $hostfound = 0;
                    $addhost   = '$newhostname';

                    open( HOSTSOURCE, '<', $hostsource ) or die "cannot open $source";
                    open( HOSTTEMP,   '>', $hosttemp )   or die "cannot open $dest";

                    while (<HOSTSOURCE>) {
                        s/HOSTNAME=\S*/HOSTNAME=$newhostname/g;
                        print HOSTTEMP $_;

                    }

                    close HOSTSOURCE;
                    close HOSTTEMP;

                    rename $hostsource => $hostbak;
                    rename $hosttemp   => $hostsource;
                    chomp( my $hostname   = `hostname` );
                    chomp( my $hostconfig = `awk '/HOSTNAME=/' /etc/sysconfig/network` );
                    my $curhost = "HOSTNAME=$newhostname";

                    if ( $hostconfig eq $curhost && $hostname eq $newhostname ) {
                        print "\t \\_ Hostname was succesfully changed from $oldhostname to $newhostname\n";
                    }

                }

                elsif ($rhel7) {
                    system("hostnamectl set-hostname $newhostname");
                    print "\t \\_ Hostname was succesfully changed from $oldhostname to $newhostname\n";
                }

                else {
                    print "\t \\_ The hostname does not appear to have been changed succesfully\n";
                    print "\t \\_ To manually fix this run: hostnamectl set-hostname newhostname.domain.com\n";
                    print "\t \\_ For example a FQDN should look like server.domain.com or server.domain.co.uk\n";
                }
            }
            else {
                print "\t \\_ To manually fix this run: hostnamectl set-hostname newhostname.domain.com\n";
                print "\t \\_ For example a FQDN should look like server.domain.com or server.domain.co.uk\n";
            }
        }
    }

    if (`dig`) {
        chomp( my $hostname  = `hostname` );
        chomp( my $digresult = `dig $hostname +short` );
        if ( $digresult eq $hostnameip ) {
            print "${green}[PASS] * The IP the hostname resolves to is the same as what's set on the server${NC}\n";
            $digpass = "yes";
        }
        elsif ( $digresult eq "" ) {
            print "${yellow}[WARN] * The hostname on the server does not resolve to an IP address${NC}\n";
            if ( $fixit == 1 ) {
                print "\t \\_ To fix this add an A record to DNS for the hostname and point it to the servers primary IP $hostnameip\n";
            }
        }
        else {
            print "${yellow}[WARN] * The hostname resolves to a different IP than what's set on the server${NC}\n";
            print "\t \\_ The hostname should resolve to $hostnameip, but actually resolves to $digresult${NC}\n";
            if ( $fixit == 1 ) {
                print "\t \\_ To fix this edit the A record in DNS for the hostname and point it to this servers primary IP $hostnameip\n";
            }
        }

    }
    else {
        print "${yellow}[WARN] * Can't check if hostname resolves to an IP because Dig is not installed.${NC}\n";
        $failed = "yes";
        if ( $fixit == 1 ) {
            if ( !$digpass ) {
                print "Would you like to fix this? Y or N: ";
                my $answer = <STDIN>;
                chomp($answer);
                if ( $answer =~ m/^y$|^yes$/i ) {
                    system("yum -y install bind-utils > /dev/null");
                    if (`dig`) {
                        print "\t \\_ dig was installed succesfully\n";
                    }
                    else {
                        print "\t \\_ dig does not appear to have been installed succesfully. Try a manual install\n";
                        print "\t \\_ To manually install dig run: yum install bind-utils\n";
                    }
                    chomp( my $hostname  = `hostname` );
                    chomp( my $digresult = `dig $hostname +short` );
                    if ( $digresult eq $hostnameip ) {
                        print "${green}[PASS] * The IP the hostname resolves to is the same as what's set on the server${NC}\n";
                        $digpass = "yes";
                    }
                    elsif ( $digresult eq "" ) {
                        print "${yellow}[WARN] * The hostname on the server does not resolve to an IP address${NC}\n";
                        print "\t \\_ To fix this add an A record to DNS for the hostname and point it to the servers primary IP $hostnameip\n";
                    }
                    else {
                        print "${yellow}[WARN] * The hostname resolves to a different IP than what's set on the server${NC}\n";
                        print "\t \\_ The hostname should resolve to $hostnameip, but actually resolves to $digresult${NC}\n";
                        print "\t \\_ To fix this edit the A record in DNS for the hostname and point it to this servers primary IP $hostnameip\n";
                    }
                }
                else {
                    print "\t \\_ To manually install dig run: yum install bind-utils\n";
                }
            }
        }
    }
}

sub oskernelchk {
    print "\n${UL}OS & KERNEL CHECKS:${UE}\n";
    if ( $rhel6 || $rhel5 ) {
        print "${green}[PASS] * The OS is Supported${NC}\n";
        print "\t \\_ $rhel\n";
    }
    elsif ($rhel7) {
        print "${green}[PASS] * The OS is Supported${NC}\n";
        print "\t \\_ $rhel7prnt\n";
    }
    elsif ($amazon) {
        print "${green}[PASS] * The OS is Supported${NC}\n";
        print "\t \\_ Amazon Linux\n";
    }
    else {
        print "${red}[FAIL] * The OS is Not Supported${NC}\n";
    }

    chomp( my $runlevel = `runlevel | awk '{print \$2}'` );
    if ( $runlevel eq 3 ) {
        print "${green}[PASS] * OS Run Level is 3${NC}\n";
    }
    else {
        print "${red}[FAIL] * OS Run Level is $runlevel${NC}\n";
        if ( $fixit == 1 ) {
            print "\t \\_ To fix this run: init 3\n";
        }
    }

    chomp( my $tmp = `stat -c "%a %n" /tmp | awk '{print \$1}'` );
    if ( $tmp eq 1777 ) {
        print "${green}[PASS] * /tmp is set to 1777 permissions${NC}\n";
    }
    else {
        print "${red}[FAIL] * /tmp is set to $tmp${NC}\n";
        if ( $fixit == 1 ) {
            print "\t \\_ To fix this run: chmod 1777 /tmp\n";
        }
    }

    chomp( my $mem = `awk '/MemTotal/ { print \$2 }' /proc/meminfo` );
    if ( $mem > 524288 ) {
        print "${green}[PASS] * System Memory Higher than 512MB${NC}\n";
    }
    else {
        print "${red}[FAIL] * System Memory Lower than 512MB${NC}\n";
        print "Please install at least 512MB to 1GB of Memory before using cPanel\n";
    }

    #change warning messages
    chomp( my $uname = `uname -r` );
    if (`uname -r | grep "grs"`) {
        print "${yellow}[WARN] * Kernel Not Supported${NC}\n";
        print "\t \\_ GRSEC Kernels are Not Supported. If you have issues installing cPanel please switch to the stock kernel and try re-installing.\n";
        print "\t \\_ $uname\n";
    }
    elsif (`uname -r | grep -i "xx"`) {
        print "${yellow}[WARN] * Kernel Not Supported${NC}\n";
        print "\t \\_ GRSEC Kernels are Not Supported. If you have issues installing cPanel please switch to the stock kernel and try re-installing.\n";
        print "\t \\_ $uname\n";
    }
    elsif (`uname -r | grep -P "2.[0-9]."`) {
        print "${green}[PASS] * Kernel Supported${NC}\n";
        print "\t \\_ $uname\n";
    }
    elsif ($rhel7) {
        if (`uname -r | grep -P "3.[0-9]."`) {
            print "${green}[PASS] * Kernel Supported as we are running $rhel7prnt${NC}\n";
            print "\t \\_ $uname\n";
        }
    }
    elsif ($amazon) {
        if (`uname -r | grep -P "4.[0-9]."`) {
            print "${green}[PASS] * Kernel Supported as we are running Amazon Linux${NC}\n";
            print "\t \\_ $uname\n";
        }
    }
    else {
        print "${yellow}[WARN] * Kernel Not Supported. If you have issues installing cPanel please switch to the stock kernel and try re-installing.${NC}\n";
        print "\t \\_ $uname\n";
    }

    print "${cyan}[INFO] * Verify System has enough Disk Space to install cPanel. You should have at least 6GB of space in / or /usr${NC}\n";
    my $dflines = `df -h`;

    foreach $line ( split /\n/, $dflines ) {
        print "\t \\_ $line\n";
    }
}
